Add support for Elixir 1.3+ Calendar types

fishcakez opened the pull request:
Initial work on Calendar types. Closes #192.

- [x] configuration
- [x] docs
- [x] timestamptz before 1970 (being handled in Elixir see https://github.com/elixir-lang/elixir/issues/5124)
- [x] use a single extension
- [x] support timetz

/cc @lau

fishcakez commented:
The test failure is unrelated and should be avoided by https://github.com/elixir-ecto/postgrex/commit/485a0a9960f962d76cbd090deae085de34616f31

josevalim commented:
@fishcakez should we put everything into a single module so it is easier to use it as an extension?

fishcakez commented:
We moved extensions into many modules a while ago because BEAM is exceptionally good at decoding integers and binaries from binaries and it was hurting performance to match on the TypeInfo struct for a lot of basic types. However for these Calendar types there is a lot more going on so likely it will be fine to use a single extension. I will do a benchmark to check once we have settled on the individual encode/decode implementations.

fishcakez commented:
@lau ping

lau commented:
@fishcakez I think it looks good.

Though I agree that it would be a lot nicer if you only had one module to pass as an extension. If there is a performance reason to having smaller modules, couldn't that be solved while still having the API be the following?:

```
start_link([database: "postgrex_test", backoff_type: :stop,
extensions: [{Postgrex.Extensions.Calendar, []}]])
```

I imagine that the code that receives the extension module names could support an "aggregate module" (e.g. Postgrex.Extensions.Calendar)  which referred to sub modules (e.g. Postgrex.Extensions.Calendar.Time).

fishcakez commented:
Added docs and made it a single extension. Still need to handle timestamptz before 1970.

fishcakez commented:
Should we also support `timetz` with `%Time{}` like #207?

lau commented:
The Elixir `DateTime.from_unix/2` function should probably support negative integers. And the `to_unix` function for that matter. Unix time integers are signed and thus can be negative. Perhaps the lowest allowed number should be -62167219200 which is "0000-01-01T00:00:00Z" or 0 Gregorian seconds. The lowest supported number in Erlang's calendar module. Maybe the simplest solution is to do the change in Elixir and have it in the next point release. /cc @josevalim

fishcakez commented:
Postgresql allows timestamps from 4713 BC to 294276 AD and dates from 4713BC to 5874897 AD so should we support all of those in postgrex?

@lau what is a point release, minor (1.4.0) or patch (1.3.3)? I would prefer to let Elixir handle it but maybe we will need to handle more dates anyway so it won't matter.

lau commented:
@fishcakez by point release I meant patch (1.3.3). To me it seems OK that people cannot use 1.3.0-1.3.2 to use datetimes from before 1970.

I should have written that the "erlang epoch" January 1st year 0, 00:00:00 UTC is the earliest time supported by Elixir `DateTime` as well. So if there is a BC datetime in the database we can't make an equivalent `DateTime` struct. `Postgrex.Timestamp` does not support negative years either.

fishcakez commented:
@lau thanks for clarifying

On 10 August 2016 at 22:28, Lau Taarnskov <notifications@github.com> wrote:

> @fishcakez <https://github.com/fishcakez> by point release I meant patch
> (1.3.3). To me it seems OK that people cannot use 1.3.0-1.3.2 to use
> datetimes from before 1970.
>
> I should have written that the "erlang epoch" January 1st year 0, 00:00:00
> UTC is the earliest time supported by Elixir DateTime as well. So if
> there is a BC datetime in the database we can't make an equivalent
> DateTime struct. Postgrex.Timestamp does not support negative years
> either.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-ecto/postgrex/pull/225#issuecomment-239010386>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6JTceDRGAJtGFj6WxyfF9ieE2iSOCAks5qekKIgaJpZM4Jb1B4>
> .
>

josevalim commented:
Let's add it to Elixir v1.3.3 then.

fishcakez commented:
rebased and added timetz support. Note that the implementation is different to #207. We will need to update master if/when this method is correct. Master will crash if the timezone makes the microseconds negative, e.g. `00:00:00+1`.

fishcakez commented:
@lau please could I get a review on the last commit

lau commented on lib/postgrex/extensions/calendar.ex:
```diff
@@ -132,25 +134,64 @@ if Code.ensure_loaded?(Calendar) do
+
+    defp timetz_to_microseconds(microsec, tz) do
+      case microsec + tz * 1_000_000 do
+        microsec when microsec < 0 ->
```
`microsec < 0`. Can microsec be negative?

lau replied:
Oh, microsec is rebound to microsec + tz and tz can make it negative, right? But `microsec` that comes as an argument is never negative. Maybe it would be clearer to use a new name for the result of microsec + tz.

fishcakez replied:
updated

fishcakez commented on test/calendar_test.exs:
```diff
@@ -0,0 +1,210 @@
+    test "decode timetz", context do
+      assert [[~T[00:00:00.000000]]] = query("SELECT time with time zone '00:00:00 UTC'", [])
+      assert [[~T[01:02:03.000000]]] = query("SELECT time with time zone '01:02:03 UTC'", [])
+      assert [[~T[23:00:00.000000]]] = query("SELECT time with time zone '00:00:00+1'", [])
```
@lau here is an example of a negative `microsec` as we need to go back 1 hour from `0 microseconds` to get UTC.

lau commented on test/calendar_test.exs:
```diff
@@ -22,6 +22,30 @@ if Code.ensure_loaded?(Calendar) do
+
+      assert :ok = query("SET SESSION TIME ZONE +1", [])
+
+      assert [[~T[12:05:06.000000]]] = query("SELECT time with time zone '04:05:06 PST'", [])
```
I like that this example is copied from before the `SET SESSION TIME ZONE +1` line.

Even though PST is one thing in postgres, outside of postgres "PST" is ambiguous, so I would add another non-UTC example that uses a numeric offset instead of a three letter code. For instance a line like this both before and after the `SET SESSION TIME ZONE +1` line:
`assert [[~T[12:05:06.000000]]] = query("SELECT time with time zone '04:05:06-8'", [])`

fishcakez replied:
Shall we replace PST with -8 elsewhere in the test suite too?

lau replied:
Yes, good idea.

fishcakez replied:
I've added a -8 test in most places we test PST

lau commented:
I think it looks good 👍

fishcakez commented:
@lau we can also go to the next day so have added handling/tests for that too for Postgrex.Time and Time.
